From 00073b412c0a87726f7ac7047d7d479f750dd73f Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:38:16 -0400
Subject: [ethernet] i40e: add missing VSI statistics

Message-id: <1409658042-22689-136-git-send-email-sassmann@redhat.com>
Patchwork-id: 90328
O-Subject: [RHEL7.1 PATCH 135/281] i40e: add missing VSI statistics
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shannon Nelson <shannon.nelson@intel.com>

Add a couple more statistics that the hardware offers but aren't part
of the standard netdev stats.

Change-ID: I201db2898f2c284aee3d9631470bc5edd349e9a5
Signed-off-by: Shannon Nelson <shannon.nelson@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 41a9e55c89be851dfc2aaae58881371b8df5360a)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index 205b559..74ecc91 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -62,6 +62,12 @@ static const struct i40e_stats i40e_gstrings_net_stats[] = {
  I40E_NETDEV_STAT(rx_crc_errors),
 };
 
+static const struct i40e_stats i40e_gstrings_misc_stats[] = {
+ I40E_VSI_STAT("rx_unknown_protocol", eth_stats.rx_unknown_protocol),
+ I40E_VSI_STAT("rx_broadcast", eth_stats.rx_broadcast),
+ I40E_VSI_STAT("tx_broadcast", eth_stats.tx_broadcast),
+};
+
 static int i40e_add_fdir_ethtool(struct i40e_vsi *vsi,
      struct ethtool_rxnfc *cmd);
 
@@ -125,7 +131,9 @@ static struct i40e_stats i40e_gstrings_stats[] = {
      * (sizeof(struct i40e_queue_stats) / sizeof(u64)))
 #define I40E_GLOBAL_STATS_LEN ARRAY_SIZE(i40e_gstrings_stats)
 #define I40E_NETDEV_STATS_LEN   ARRAY_SIZE(i40e_gstrings_net_stats)
+#define I40E_MISC_STATS_LEN ARRAY_SIZE(i40e_gstrings_misc_stats)
 #define I40E_VSI_STATS_LEN(n)   (I40E_NETDEV_STATS_LEN + \
+     I40E_MISC_STATS_LEN + \
      I40E_QUEUE_STATS_LEN((n)))
 #define I40E_PFC_STATS_LEN ( \
   (FIELD_SIZEOF(struct i40e_pf, stats.priority_xoff_rx) + \
@@ -648,6 +656,11 @@ static void i40e_get_ethtool_stats(struct net_device *netdev,
   data[i++] = (i40e_gstrings_net_stats[j].sizeof_stat ==
    sizeof(u64)) ? *(u64 *)p : *(u32 *)p;
  }
+ for (j = 0; j < I40E_MISC_STATS_LEN; j++) {
+  p = (char *)vsi + i40e_gstrings_misc_stats[j].stat_offset;
+  data[i++] = (i40e_gstrings_misc_stats[j].sizeof_stat ==
+       sizeof(u64)) ? *(u64 *)p : *(u32 *)p;
+ }
  rcu_read_lock();
  for (j = 0; j < vsi->num_queue_pairs; j++) {
   tx_ring = ACCESS_ONCE(vsi->tx_rings[j]);
@@ -714,6 +727,11 @@ static void i40e_get_strings(struct net_device *netdev, u32 stringset,
      i40e_gstrings_net_stats[i].stat_string);
    p += ETH_GSTRING_LEN;
   }
+  for (i = 0; i < I40E_MISC_STATS_LEN; i++) {
+   snprintf(p, ETH_GSTRING_LEN, "%s",
+     i40e_gstrings_misc_stats[i].stat_string);
+   p += ETH_GSTRING_LEN;
+  }
   for (i = 0; i < vsi->num_queue_pairs; i++) {
    snprintf(p, ETH_GSTRING_LEN, "tx-%u.tx_packets", i);
    p += ETH_GSTRING_LEN;
diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 93ec7ea..cf14f66 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -530,6 +530,12 @@ void i40e_update_eth_stats(struct i40e_vsi *vsi)
  i40e_stat_update32(hw, I40E_GLV_RDPC(stat_idx),
       vsi->stat_offsets_loaded,
       &oes->rx_discards, &es->rx_discards);
+ i40e_stat_update32(hw, I40E_GLV_RUPP(stat_idx),
+      vsi->stat_offsets_loaded,
+      &oes->rx_unknown_protocol, &es->rx_unknown_protocol);
+ i40e_stat_update32(hw, I40E_GLV_TEPC(stat_idx),
+      vsi->stat_offsets_loaded,
+      &oes->tx_errors, &es->tx_errors);
 
  i40e_stat_update48(hw, I40E_GLV_GORCH(stat_idx),
       I40E_GLV_GORCL(stat_idx),
@@ -807,6 +813,8 @@ static void i40e_update_vsi_stats(struct i40e_vsi *vsi)
  ns->tx_errors = es->tx_errors;
  ons->multicast = oes->rx_multicast;
  ns->multicast = es->rx_multicast;
+ ons->rx_dropped = oes->rx_discards;
+ ns->rx_dropped = es->rx_discards;
  ons->tx_dropped = oes->tx_discards;
  ns->tx_dropped = es->tx_discards;
 
-- 
1.7.1